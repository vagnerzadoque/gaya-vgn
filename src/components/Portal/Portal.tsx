import { ReactPortal, useEffect, useState } from 'react'
import { createPortal } from 'react-dom'

export interface PortalProps {
  children: React.ReactNode
  container?: Element | (() => Element | null) | null
}

export const Portal = ({ children, container }: PortalProps): ReactPortal | null => {
  const [mountNode, setMountNode] = useState<Element | null>()

  useEffect(() => {
    const element = container || document.body
    return setMountNode(element)
  }, [container])

  return mountNode ? createPortal(children, mountNode) : null
}

export default Portal
