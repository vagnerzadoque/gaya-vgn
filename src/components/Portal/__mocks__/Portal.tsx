import React from 'react'
import { PortalProps } from '../Portal'

export const Portal = ({ children }: PortalProps): JSX.Element => (<div data-testid="ds-portal">{children}</div>)

export default Portal
