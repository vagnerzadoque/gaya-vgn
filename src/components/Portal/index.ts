import Portal from './Portal'

export { Portal }
export { PortalProps } from './Portal'
export default { Portal }
