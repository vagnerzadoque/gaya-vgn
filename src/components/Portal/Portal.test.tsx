import React from 'react'
import { render } from '@testing-library/react'
import { Portal } from './Portal'

describe('Portal', () => {
  beforeAll(() => {
    document.body.innerHTML = '<div id="container"></div>'
  })
  it('should render element into body as default', () => {
    const component = render(<Portal><p data-testid="something">something</p></Portal>)

    expect(component.baseElement).toMatchSnapshot()
    expect(document.body).toContainElement(component.getByTestId('something'))
  })

  it('should render element into the given container', () => {
    const component = render(<Portal container={() => document.getElementById('container')}><p data-testid="something">something</p></Portal>)

    expect(component.baseElement).toMatchSnapshot()
    expect(document.getElementById('container')).toContainElement(component.getByTestId('something'))
  })
})
