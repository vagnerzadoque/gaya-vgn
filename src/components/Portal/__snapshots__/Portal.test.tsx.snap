// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Portal should render element into body as default 1`] = `
<body>
  <div
    id="container"
  />
  <div />
  <p
    data-testid="something"
  >
    something
  </p>
</body>
`;

exports[`Portal should render element into the given container 1`] = `
<body>
  <div
    id="container"
  >
    <p
      data-testid="something"
    >
      something
    </p>
  </div>
  <div />
</body>
`;
